[contrib] tool to run Boost Test based C++ unittests in parallel

ftrader opened the pull request:
contrib/testtools/gtest-parallel-bitcoin is a rough adaptation of Google's test sharding wrapper at https://github.com/google/gtest-parallel . The license is Apache 2.0 (see file header).

It has been adapted to use the Boost Test framework's '--list_content' command line parameter, and some Google Test dependencies have been removed.

It can be run by simply giving it the path to the test_bitcoin executable, which it then queries to obtain a list of runnable tests.
This only works if Bitcoin has been compiled against Boost 1.59 or later.
On earlier versions, the executable would unfortunately ignore the '--list_content' option and start running all the tests it knows.

Running with `--help` provides help about the command line options.
Using `--format=raw` option provides detailed worker thread output during execution.
The `--print_test_times` will print a detailed breakdown of all test times.

The parallelization wrapper creates and maintains a score database of test execution times in ~/.gtest-parallel-bitcoin-times .
When it executes tests in repeated runs, it starts them off in order from longest to shortest test to save time. This is different from `-r` option which can be used for repeating tests within a run.

You can specify the `--failed` option to run only previously failed and new tests.
The `--timeout` option can be used to constrain runtime.
Interrupted tests will be listed.

Testing on a desktop machine with sufficient cores shows that the total test execution time when using this parallelization script is is reduced to approximately the running time of the slowest unit test.

This tool is an updated, pep8-formatted version based on my original adaptation at https://github.com/ftrader-bitcoinunlimited/hardfork_prototype_1_mvf-bu/commit/c757a8299e9073910ad5e09e94d14d55a89ee865

-----------

Example usage (once build has been completed):

    contrib/testtools/gtest-parallel-bitcoin src/test/test_bitcoin

Example output:

```
$ contrib/testtools/gtest-parallel-bitcoin src/test/test_bitcoin
[13/54] scheduler_tests (21ms)
Test setup error: no test cases matching filter or all test cases were disabled
[13/54] scheduler_tests  returned/aborted with exit code 200 (21 ms)
[54/54] PrevectorTests (54181ms)
FAILED TESTS (1/54):
src/test/test_bitcoin: scheduler_tests
```

ftrader committed:
[contrib] tool to run Boost Test based C++ unittests in parallel

contrib/testtools/gtest-parallel-bitcoin is a rough adaptation of Google's test sharding wrapper at https://github.com/google/gtest-parallel .
The license is Apache 2.0 (see file header).

It has been adapted to use the Boost Test framework's '--list_content' command line parameter, and requires Boost version 1.59 or later.
Some Google Test dependencies have been removed.

With enough free cores, the total test execution time when using this parallelization script can be reduced approximately to the running time of the slowest unit test.

This tool is an updated, pep8-formatted version based on my original adaptation at https://github.com/ftrader-bitcoinunlimited/hardfork_prototype_1_mvf-bu/commit/c757a8299e9073910ad5e09e94d14d55a89ee865

Example usage:

    contrib/testtools/gtest-parallel-bitcoin src/test/test_bitcoin

zander reviewed (commented):
I like the idea, thanks!

I tried it on master and this is the output of a run;
```
[11/54] scheduler_tests  (34 ms)                                                                                                    
Test setup error: no test cases matching filter or all test cases were disabled
[11/54] scheduler_tests  returned/aborted with exit code 200 (34 ms)
[54/54] PrevectorTests (51235 ms)                                                                                                   
FAILED TESTS (1/54):
 src/test/test_bitcoin: scheduler_tests 
```

When I ran the test normally, it finished without errors.

I also noticed that the test doesn't actually print what it is running currently. This is easy to see as the prevector test takes several seconds, but the 'miner' test is shown to take several seconds.

(I'm using boots 1.63)

zander commented:
Great, github lost my other review comment;

your new file attributes copyright to "Bitcoin developers". This is not a legal entity (1) and as such can't be used to own copyright.
Please replace them with actual persons names and email addresses. Thanks!

1) http://thelawdictionary.org/legal-entity

ftrader commented:
```
./policy/fees.cpp:2:// Copyright (c) 2009-2015 The Bitcoin developers
./policy/policy.cpp:2:// Copyright (c) 2009-2015 The Bitcoin developers
./policy/fees.h:2:// Copyright (c) 2009-2015 The Bitcoin developers
./policy/policy.h:2:// Copyright (c) 2009-2015 The Bitcoin developers
./memusage.h:1:// Copyright (c) 2015 The Bitcoin developers
./arith_uint256.h:2:// Copyright (c) 2009-2015 The Bitcoin developers
./arith_uint256.cpp:2:// Copyright (c) 2009-2014 The Bitcoin developers
./qt/paymentrequestplus.h:1:// Copyright (c) 2011-2015 The Bitcoin developers
./compat/byteswap.h:1:// Copyright (c) 2014 The Bitcoin developers
./compat/endian.h:1:// Copyright (c) 2014-2015 The Bitcoin developers
```

Most recent merge featuring such an attribution: `policy.h`, 877fc1351db93f3aee80340deb95b00b20700c29

What's your impression of the practice from 2009-2015 (several files)?

The `scheduler_test` issue is reported correctly by the script ("no test cases matching filter or all test cases were disabled"). No idea why Boost ignores this silently when running the suite as-is.

It is of course due to the only test case being commented out (dead code).
```
#if 0 /* Disabled for now because there is a race condition issue in this test - see #6540 */
BOOST_AUTO_TEST_CASE(manythreads)
```

zander commented:
> Most recent merge featuring such an attribution: policy.h, 877fc13

That merge doesn't add enough copyrightable material to claim.

I am aware that most files in Bitcoin have this problem that they assign copyright to a non-existing entity. Which is a shame, but I can't change the past. I frankly expected you to be happy that you can assign the copyright to yourself instead of an anonymous group like "Bitcoin developers"...  What is the reason you send a reply instead of change your file?

>  No idea why Boost ignores this silently when running the suite as-is.

The problem is that it doesn't ignore it, the problem is that it fails :)

> It is of course due to the only test case being commented out (dead code).

Agreed, that seems like something that should be updated or removed. Its lucky that it caught a bug in the script, though! Would you not agree?

Maybe you can make it so that should there be a suite but not tests, it is not actually ran?

Are you willing to fix the copyright lines at least, please?

zander commented:
For ease of usage, here is the boost output of the unit test as it is run by the python script; [test-list-content.txt](https://github.com/bitcoinclassic/bitcoinclassic/files/930796/test-list-content.txt)

ftrader commented:
Would 'The Bitcoin Classic developers' be a valid legal entity?

zander commented:
> Would 'The Bitcoin Classic developers' be a valid legal entity?

No, it would not be.  "Google Inc" is, My full name plus email addy is.
Freetrader as a name combined with your email address is (although I prefer a legal name, I understand you like to keep an alias, and thats fine).

ftrader committed:
Elaborate on copyright for the changes made to the script (if it doesn't add enough then I waive the claim)

ftrader commented:
Thanks for the clarification. I've adapted the copyright line.
Why I asked about the 'Bitcoin Classic developers':

```
init.cpp:312:           FormatParagraph(strprintf("Copyright (C) 2016-%i The Bitcoin Classic developers", COPYRIGHT_YEAR)) + "\n" +
```

ftrader commented:
The patched `scheduler_tests.cpp` from BU (changes merged from Core 0.12 I believe) works fine for me btw (when I run it on Classic). The diff is pretty minimal, essentially only this (and the removal of the `#ifdef 0`:

```
 static void MicroSleep(uint64_t n)
 {
+#if defined(HAVE_WORKING_BOOST_SLEEP_FOR)
     boost::this_thread::sleep_for(boost::chrono::microseconds(n));
+#elif defined(HAVE_WORKING_BOOST_SLEEP)
+    boost::this_thread::sleep(boost::posix_time::microseconds(n));
+#else
+    //should never get here
+    #error missing boost sleep implementation
+#endif
 }
```

If you'd like to I can submit a separate PR for that.

zander commented:
The copyright string in the init.cpp is not a legal statement, it is a user-visible string that helps users understand. It is a little odd as we depend on lots of dependencies which are not included in the copyright claim. In most software you would see the license listed instead.

As you may know, I have spend many years on many open source projects, and Bitcoin as we inherited it has many weird habits. But I don't think changing all at once is healthy either ;)

Thanks for your copyright line fix!

Do you think you can fix the bug in this script failing? I'll fix the unit test, but that will just hide the bug, assuming we never end up with another empty unit test in the future...

ftrader committed:
Treat empty test suites as PASSED by default (flag them with --fail_empty_suites)

Previous behaviour was to raise an error ("no test cases matching filter or all test cases were disabled") when running a test suite which contained no test cases.
This reflected Boost Test returning exit code 200 for such a suite.

ftrader commented:
I changed the behavior of the program to treat empty test suites as PASSED by default.
While I'm not sure this is an entirely sensible default, it can be overridden by specifying `--fail_empty_suites` which will show up the previous error. I have tested with and without the flag.
With the flag (and the unpatched scheduler_tests.cpp, the script returns exit code 200, as before). Without it, it now declares all 54 unit tests passed and exits with code 0.

I am not a lawyer but I always assumed that if a software shows me an explicit copyright message, then it must be trying to convey some legally significant message through that statement, otherwise why make it. One could as well write on it "brought to you by the fine folks at Bitcoin Classic" or "for details on copyright, read the individual source files". That might actually encourage people to read the source!

zander commented:
> I am not a lawyer but I always assumed that if a software shows me an explicit copyright message, then it must be trying to convey some legally significant message through that statement, otherwise why make it. One could as well write on it "brought to you by the fine folks at Bitcoin Classic" or "for details on copyright, read the individual source files". That might actually encourage people to read the source!

I'm certainly not going to disagree with that! :)

If you look at many pieces of open source software, you'll find your thoughts to be in line with their practices. Where a copyright line is shown its for things like the Free Software Foundation or otherwise actually existing legal entities.  
This requires copyright assignment and a foundation, neither of which is appropriate for Bitcoin. (more;  https://www.gnu.org/licenses/why-assign.html )

I'm not sure if its politically a good idea to remove the "copyright" line from the splashscreen or from the command line help right now. Its legally completely irrelevant, though.

thanks for the contrib! I merged it :)

zander commented:
The last issue I mentioned in my initial review, I fixed that in 4be660756
